docs/source/projects.rst

```diff
@@ -271,6 +271,63 @@ Use cases for the **job document** include, but are not limited to:
   2) Tracking of **runtime information**
   3) **labeling** of jobs, e.g. to identify error states.
 
+.. tip::
+
+    Large arrays of numerical data are often not conducive to store in the :py:attr:`Job.document`. Text-based formats like JSON can be slower and less precise for floating-point numbers. For this kind of data, consider using the :py:attr:`Job.data` attribute.
+
+.. _project-job-data:
+
+Job Data Storage
+----------------
+
+Large numerical or text data can be stored in :py:attr:`Job.data`.
+This uses a file in `HDF5`_ format to store array-like or dictionary-like information.
+Like the :py:attr:`Job.document`, this information can be accessed using key-value pairs.
+Unlike the :py:attr:`Job.document`, job data is not searchable.
+
+.. _`HDF5`: https://portal.hdfgroup.org/display/HDF5/HDF5
+
+An example of storing data:
+
+.. code-block:: python
+
+    >>> import numpy as np
+    >>> job = project.open_job(statepoint)
+    >>> job.data['x'] = np.arange(0, 1, 0.01)
+    >>> job.data['hello'] = 'world'
+
+
+Just like the job *state point* and *document*, individual keys may be accessed either as attributes or through a functional interface, *e.g.*.
+The following examples are all equivalent:
+
+.. code-block:: python
+
+    >>> print(job.data.get('hello'))
+    world
+    >>> print(job.data['hello'])
+    world
+    >>> print(job.data.hello)
+    world
+
+The underlying HDF5-file is openend and flushed after each read- and write-operation.
+You can keep the file explicitily open using a context manager.
+The file is only opened and flushed once in the following example:
+
+.. code-block:: python
+
+    >>> with job.data:
+    ...     job.data['hello'] = 'world'
+    ...     print(job.data.x)
+    ...
+
+.. tip::
+
+     Use the :py:meth:`Job.data.get` method to return ``None`` or another specified default value for missing values. This works exactly like with python's `built-in dictionaries <https://docs.python.org/3/library/stdtypes.html#dict.get>`_.
+
+.. warning::
+
+    It is strongly advised that operations on :py:attr:`Job.data` are not performed in parallel, to avoid data corruption.
```

Comment:
I think we should clarify that HDF5 libraries usually do not permit concurrent access. There is no risk of data corruption. 

Maybe add a link to a relevant section as part of the h5py or the HDF5 documentation.

Reply:
Are you sure there is no risk of data corruption even if you concurrently write to / delete from the same HDF5 file on a network file system from two different computers? Even parallel access from multiple processes on the same computer is not guaranteed to be safe (exceptions: thread-safe, MPI-safe implementations of Parallel HDF5, which do not apply to us). This said, I have not tested it myself - I've only read about the dangers.

Related:
https://stackoverflow.com/questions/35837243/hdf5-possible-data-corruption-or-loss
https://news.ycombinator.com/item?id=10860496

Reply:
I didn't see anything about data corruption through parallel access and so far have not experienced any such issues. However, data corruption may occur in cases where an application crashed. That said, I'm ok leaving the warning, maybe it should be a bit more differentiated. 

However, reading all of this makes me wonder if we should return to my original design of having *one file per top-level key*. That would be much safer, would allow concurrent access to the top-level keys, avoid the whole repacking issue, and would also significantly simplify synchronization. What do you think?